File: app/schema/girl_companion.py
"""Structured contracts for the GirlAI companion turn."""

from datetime import datetime
from typing import Dict, List, Literal, Optional

from pydantic import BaseModel, Field, field_validator


EmotionLabel = Literal[
    "neutral",
    "happy",
    "sad",
    "anxious",
    "stressed",
    "tired",
    "angry",
    "overwhelmed",
    "focused",
]
IntentLabel = Literal[
    "unknown",
    "chat",
    "acknowledge",
    "task_planning",
    "task_execution",
    "task_review",
    "task_blocked",
    "rest_request",
    "help_request",
    "remember_preference",
]


class CompanionTurnRequest(BaseModel):
    """Request body for a structured companion turn."""

    prompt: str = Field(..., min_length=1, max_length=2000)
    character_id: str = Field(default="gentle", min_length=1, max_length=128)
    turn_id: Optional[str] = Field(default=None, min_length=1, max_length=128)
    temperature: Optional[float] = Field(default=None, ge=0.0, le=1.5)
    max_tokens: Optional[int] = Field(default=None, ge=50, le=1000)
    voice_output: bool = False


class VoiceTranscriptionRequest(BaseModel):
    """Standardized transcription result accepted by the voice adapter API."""

    transcript: str = Field(..., min_length=1, max_length=2000)
    character_id: str = Field(default="gentle", min_length=1, max_length=128)
    turn_id: Optional[str] = Field(default=None, min_length=1, max_length=128)
    provider: Optional[str] = Field(default=None, max_length=128)
    confidence: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    duration_ms: Optional[int] = Field(default=None, ge=0, le=3_600_000)
    voice_output: bool = False

    @field_validator("transcript")
    @classmethod
    def validate_transcript(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("transcript must contain visible text")
        return value


class EmotionState(BaseModel):
    """Normalized emotion state returned by the companion pipeline."""

    label: EmotionLabel = "neutral"
    intensity: float = Field(default=0.0, ge=0.0, le=1.0)
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    raw_label: Optional[str] = Field(default=None, max_length=64)
    low_confidence: bool = False


class IntentState(BaseModel):
    """Normalized work intent returned by the companion pipeline."""

    label: IntentLabel = "unknown"
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    raw_label: Optional[str] = Field(default=None, max_length=64)
    low_confidence: bool = False


class MemoryCandidate(BaseModel):
    """A memory suggestion awaiting explicit user confirmation."""

    id: Optional[str] = Field(default=None, max_length=36)
    key: str = Field(..., min_length=1, max_length=50)
    value: str = Field(..., min_length=1, max_length=2000)
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    source: str = Field(default="conversation", min_length=1, max_length=64)


class CompanionMemoryConfirmRequest(BaseModel):
    """User confirmation and optional revision for a memory candidate."""

    key: Optional[str] = Field(default=None, min_length=1, max_length=50)
    value: Optional[str] = Field(default=None, min_length=1, max_length=2000)
    visibility: Literal["conversation_only", "companion_allowed"] = "companion_allowed"


class CompanionMemoryResponse(BaseModel):
    """User-owned memory record exposed by the management API."""

    id: str
    key: str
    value: str
    confidence: int = Field(ge=0, le=100)
    source: str
    status: Literal["candidate", "confirmed", "rejected", "deleted"]
    consent_source: Literal["user_confirmed", "imported", "system_derived"]
    visibility: Literal["conversation_only", "companion_allowed"]
    last_used_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class CompanionMemoryPage(BaseModel):
    """Paginated memory records for the authenticated user."""

    memories: List[CompanionMemoryResponse]
    total: int = Field(ge=0)
    limit: int = Field(ge=1)
    offset: int = Field(ge=0)


class ModelContext(BaseModel):
    """Safe model execution metadata; credentials are excluded by contract."""

    current_model: Optional[str] = Field(default=None, max_length=256)
    classification_model: Optional[str] = Field(default=None, max_length=256)
    current_agent: Optional[str] = Field(default=None, max_length=128)
    calls: int = Field(default=0, ge=0)
    fallback_used: bool = False
    fallback_history: List[str] = Field(default_factory=list)


class VoiceOutputState(BaseModel):
    """Voice output status associated with the same companion turn."""

    requested: bool = False
    status: Literal["disabled", "unavailable", "ready", "failed"] = "disabled"
    provider: Optional[str] = Field(default=None, max_length=128)
    error: Optional[str] = Field(default=None, max_length=256)


class VoiceInputState(BaseModel):
    """Normalized transcription metadata associated with a companion turn."""

    received: bool = False
    status: Literal["disabled", "received", "failed"] = "disabled"
    provider: Optional[str] = Field(default=None, max_length=128)
    confidence: Optional[float] = Field(default=None, ge=0.0, le=1.0)
    duration_ms: Optional[int] = Field(default=None, ge=0, le=3_600_000)


class CompanionTurn(BaseModel):
    """Versioned structured output for one GirlAI companion turn."""

    turn_id: Optional[str] = Field(default=None, max_length=128)
    conversation_id: Optional[str] = Field(default=None, max_length=128)
    assistant_text: str = Field(..., min_length=1, max_length=20000)
    emotion: EmotionState = Field(default_factory=EmotionState)
    intent: IntentState = Field(default_factory=IntentState)
    care_required: bool = False
    response_style: Literal["standard", "neutral", "care"] = "standard"
    work_options: List[str] = Field(default_factory=list, max_length=3)
    memory_candidates: List[MemoryCandidate] = Field(default_factory=list)
    model_context: ModelContext = Field(default_factory=ModelContext)
    voice_input: VoiceInputState = Field(default_factory=VoiceInputState)
    voice_output: VoiceOutputState = Field(default_factory=VoiceOutputState)
    degraded_capabilities: List[str] = Field(default_factory=list)
    schema_version: int = Field(default=1, ge=1)


class CompanionTurnResponse(CompanionTurn):
    """API response with persisted conversation state metadata."""

    model: str = Field(..., max_length=256)
    tokens_used: int = Field(default=0, ge=0)
    state_revision: int = Field(default=0, ge=0)

File: flutter_client/lib/domain/models/girl_companion.dart
class EmotionState {
  const EmotionState({
    this.label = 'neutral',
    this.intensity = 0,
    this.confidence = 0,
  });
  final String label;
  final double intensity;
  final double confidence;
  factory EmotionState.fromJson(Map<String, dynamic> j) => EmotionState(
    label: '${j['label'] ?? 'neutral'}',
    intensity: _d(j['intensity']),
    confidence: _d(j['confidence']),
  );
}

class IntentState {
  const IntentState({this.label = 'unknown', this.confidence = 0});
  final String label;
  final double confidence;
  factory IntentState.fromJson(Map<String, dynamic> j) => IntentState(
    label: '${j['label'] ?? 'unknown'}',
    confidence: _d(j['confidence']),
  );
}

class MemoryCandidate {
  const MemoryCandidate({
    this.id,
    this.key = '',
    this.value = '',
    this.confidence = 0,
    this.source = 'conversation',
  });
  final String? id;
  final String key, value, source;
  final double confidence;
  factory MemoryCandidate.fromJson(Map<String, dynamic> j) => MemoryCandidate(
    id: j['id']?.toString(),
    key: '${j['key'] ?? ''}',
    value: '${j['value'] ?? ''}',
    confidence: _d(j['confidence']),
    source: '${j['source'] ?? 'conversation'}',
  );
}

class VoiceState {
  const VoiceState({this.status = 'disabled', this.provider, this.confidence});
  final String status;
  final String? provider;
  final double? confidence;
  factory VoiceState.fromJson(Map<String, dynamic> j) => VoiceState(
    status: '${j['status'] ?? 'disabled'}',
    provider: j['provider']?.toString(),
    confidence: j['confidence'] == null ? null : _d(j['confidence']),
  );
}

class CompanionTurnResponse {
  const CompanionTurnResponse({
    this.text = '',
    this.turnId,
    this.conversationId,
    this.emotion = const EmotionState(),
    this.intent = const IntentState(),
    this.memories = const [],
    this.voiceInput = const VoiceState(),
    this.voiceOutput = const VoiceState(),
    this.stateRevision = 0,
  });
  final String text;
  final String? turnId, conversationId;
  final EmotionState emotion;
  final IntentState intent;
  final List<MemoryCandidate> memories;
  final VoiceState voiceInput, voiceOutput;
  final int stateRevision;
  factory CompanionTurnResponse.fromJson(Map<String, dynamic> j) =>
      CompanionTurnResponse(
        text: '${j['assistant_text'] ?? ''}',
        turnId: j['turn_id']?.toString(),
        conversationId: j['conversation_id']?.toString(),
        emotion: EmotionState.fromJson(
          Map<String, dynamic>.from(j['emotion'] ?? {}),
        ),
        intent: IntentState.fromJson(
          Map<String, dynamic>.from(j['intent'] ?? {}),
        ),
        memories: [
          for (final x in (j['memory_candidates'] as List? ?? const []))
            MemoryCandidate.fromJson(Map<String, dynamic>.from(x)),
        ],
        voiceInput: VoiceState.fromJson(
          Map<String, dynamic>.from(j['voice_input'] ?? {}),
        ),
        voiceOutput: VoiceState.fromJson(
          Map<String, dynamic>.from(j['voice_output'] ?? {}),
        ),
        stateRevision: int.tryParse('${j['state_revision'] ?? 0}') ?? 0,
      );
}

class GirlCharacter {
  const GirlCharacter({
    required this.id,
    required this.name,
    this.description = '',
    this.tags = const [],
  });
  final String id, name, description;
  final List<String> tags;
  factory GirlCharacter.fromJson(Map<String, dynamic> j) => GirlCharacter(
    id: '${j['id']}',
    name: '${j['name'] ?? j['id']}',
    description: '${j['description'] ?? ''}',
    tags: [for (final x in (j['tags'] as List? ?? const [])) '$x'],
  );
}

class GirlHistory {
  const GirlHistory({
    required this.id,
    required this.role,
    required this.content,
    this.createdAt,
  });
  final String id, role, content;
  final DateTime? createdAt;
  factory GirlHistory.fromJson(Map<String, dynamic> j) => GirlHistory(
    id: '${j['id']}',
    role: '${j['role'] ?? 'assistant'}',
    content: '${j['content'] ?? ''}',
    createdAt: DateTime.tryParse('${j['created_at'] ?? ''}'),
  );
}

class GirlPreference {
  const GirlPreference({
    required this.id,
    required this.key,
    required this.value,
  });
  final String id, key, value;
  factory GirlPreference.fromJson(Map<String, dynamic> j) => GirlPreference(
    id: '${j['id']}',
    key: '${j['key'] ?? j['preference_key'] ?? ''}',
    value: '${j['value'] ?? j['preference_value'] ?? ''}',
  );
}

double _d(dynamic v) => double.tryParse('$v') ?? 0;
